pkg/glusterfs/nodeserver.go

```diff
@@ -0,0 +1,101 @@
+/*
+Copyright 2018 The Gluster CSI Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package glusterfs
+
+import (
+	"fmt"
+	"os"
+	"strings"
+
+	"github.com/container-storage-interface/spec/lib/go/csi"
+	"golang.org/x/net/context"
+	"google.golang.org/grpc/codes"
+	"google.golang.org/grpc/status"
+	"k8s.io/kubernetes/pkg/util/mount"
+	"k8s.io/kubernetes/pkg/volume/util"
+
+	"github.com/kubernetes-csi/drivers/pkg/csi-common"
+)
+
+type nodeServer struct {
+	*csicommon.DefaultNodeServer
+}
+
+func (ns *nodeServer) NodePublishVolume(ctx context.Context, req *csi.NodePublishVolumeRequest) (*csi.NodePublishVolumeResponse, error) {
+	targetPath := req.GetTargetPath()
+	notMnt, err := mount.New("").IsLikelyNotMountPoint(targetPath)
+	if err != nil {
+		if os.IsNotExist(err) {
+			if err := os.MkdirAll(targetPath, 0750); err != nil {
+				return nil, status.Error(codes.Internal, err.Error())
+			}
+			notMnt = true
+		} else {
+			return nil, status.Error(codes.Internal, err.Error())
+		}
+	}
+
+	if !notMnt {
+		return &csi.NodePublishVolumeResponse{}, nil
+	}
+
+	mo := req.GetVolumeCapability().GetMount().GetMountFlags()
+	if req.GetReadonly() {
+		mo = append(mo, "ro")
+	}
+
+	s := req.GetVolumeAttributes()["glusterserver"]
```

Comment:
We need a list of servers, so that backup servers can be used? This relies on one server, and can hence fail?

See, https://github.com/gluster/gluster-subvol/blob/master/volcreator/creator.sh#L58

and see, https://github.com/gluster/gluster-subvol/blob/master/glfs-subvol/glfs-subvol#L87 on how this is used in the subvol flexvol provider

Reply:
These are already available in the attribute "glusterbkpservers" (see controllerserver.go:115). We should get that out here and put that into back volfile servers for mount. it's made for that.

Reply:
exactly as @obnoxxx mentioned, its already exposed in a var and I forgot to put it in the mount, will be taken care in the next commit!